### ISSUE TITLE
Bad subtitle encoding

### ISSUE THREAD

**Issue by halali** [morpheus65535/bazarr#38]
Hi

when bazarr download czech or slovak subtitles where is diacritics it show like this,

> na jejĂ­ pÄ›tiletĂ© misi.
Zkoumat neznĂˇmĂ© novĂ© svÄ›ty,

**Comment by morpheus65535** [comment]
It's strange because I explicitly ask to Subliminal module to save content of subtitles in utf-8: https://github.com/morpheus65535/bazarr/blob/97652cf0728bb8ab858c4adafb1963344a6565e0/get_subtitle.py#L38

It's been tested a lot (like a thousand time) without any problem. Do you have this problem with only one subtitles or more?

**Comment by halali** [comment]
more, all that i download...

had same problem with sub-zero and they solve it somehow...
but i dont use sub-zero anymore, it download lot of shit subtitles

**Comment by morpheus65535** [comment]
@zeusy123 You had problems with those languages?

**Comment by morpheus65535** [comment]
@halali Still have this issue? You can point me on an Sub-Zero issue where I can find more information on what they've done to fix it? Thanks!

**Comment by halali** [comment]
Yes still have it

try this 
https://github.com/pannal/Sub-Zero.bundle/commit/e274a542c116e64d9804139f0b03725705911a49
https://github.com/pannal/Sub-Zero.bundle/commit/419bee76e203ff8689bb605008f3a1e87eb93105

**Comment by morpheus65535** [comment]
Ok seems they are using their own patched version of subliminal. I would prefer not to go down this path...

I've hear that @Diaoul is working on a new major version of Subliminal (v3). I hope he'll fix that kind of stuff in it.

**Comment by morpheus65535** [comment]
@halali Could you switch to dev branch and update. You're going to be using the last development version of Subliminal. Please report back.

**Comment by halali** [comment]
yes it work now

**Comment by gszigethy** [comment]
This is a problem for me too that started recently. Attaching the original files from the subtitle sites downloaded and the subtitle file that has been processed by Bazarr. Basically when the subtitle is re-coded from ANSI to UTF-8, it displays incorrect local characters on both of my Samsung TVs as well as on Windows 10 using BSPlayer.
Is there a chance that re-coding can be made optional please with a switch? Currently this makes Bazarr unusable for me because I have to re-download all subtitles again.
Attaching a few examples. They have been downloaded from Supersubtitles (hungarian provider).
[conversations.with.a.killer.bazarr.zip](https://github.com/morpheus65535/bazarr/files/3168789/conversations.with.a.killer.bazarr.zip)
[Conversations.with.a.Killer.The.Ted.Bundy.Tapes.S01.720p.NF.WEB-DL.DD5.1.x264-QOQ.hun.zip](https://github.com/morpheus65535/bazarr/files/3168790/Conversations.with.a.Killer.The.Ted.Bundy.Tapes.S01.720p.NF.WEB-DL.DD5.1.x264-QOQ.hun.zip)
[chernobyl.zip](https://github.com/morpheus65535/bazarr/files/3168794/chernobyl.zip)

**Comment by gszigethy** [comment]
What can be done with this? All my subtitles downloaded that are re-coded from ANSI to UTF-8 are incorrect, basically Bazarr became unusable. I'm using the docker image. No local characters are coded well.

Thanks.

**Comment by morpheus65535** [comment]
@pannal could we make this optionnal?
https://github.com/morpheus65535/bazarr/blob/59ac0a8b910189c6a958270e91080aab16560e81/libs/subliminal_patch/core.py#L312

**Comment by pannal** [comment]
Converting to UTF-8 is not an issue at all. If it's done correctly.

This might be a special case for Hungarian, I'll check.

**Comment by pannal** [comment]
Checking chernobyl.zip:
Both look similar to me, for example line 12: `Ő volt a legjobb választás.` vs. `Õ volt a legjobb választás.`

Are your clients perhaps not capable of displaying UTF-8?

**Comment by gszigethy** [comment]
Thank you guys for looking into it. It is not just those characters, let me show you. It is visible with all characters.

**1) Bazarr-normalized subtitle**
This has been normalized by Bazarr and converted to UTF-8.
- BSPlayer.
![image](https://user-images.githubusercontent.com/46370269/58306125-be544300-7dfa-11e9-8d63-cd2f0f332631.png)
- Samsung TV
![image](https://user-images.githubusercontent.com/46370269/58306353-8a2d5200-7dfb-11e9-9347-b890c54e08fe.png)
- This is the subtitle:
[Chernobyl.S01E01.720p.WEBRip.x264-TBS.bazarr.zip](https://github.com/morpheus65535/bazarr/files/3215353/Chernobyl.S01E01.720p.WEBRip.x264-TBS.bazarr.zip)
As you see, two clients (and my other TV as well) interprets them the same way, incorrectly.
- Notepad++ on Windows says this subtitle is UTF-8:
![image](https://user-images.githubusercontent.com/46370269/58306454-d5dffb80-7dfb-11e9-8372-7c033d5a1a94.png)
- If I check the file in Debian:
```
root@llamaduck:/srv/tomb/jbod/data/sorozat/Chernobyl/Season 01# uchardet Chernobyl.S01E01.720p.WEBRip.x264-TBS.bazarr.srt 
UTF-8
```

**2) Original subtitle downloaded from the provider:**
This is Windows-1252 encoded.
- BSPLayer
![image](https://user-images.githubusercontent.com/46370269/58306684-9b2a9300-7dfc-11e9-90e2-8ceb8459e777.png)
- Samsung TV
![image](https://user-images.githubusercontent.com/46370269/58306754-dcbb3e00-7dfc-11e9-89f3-ee952e6f326e.png)
- This is the subtitle:
[Chernobyl.S01E01.720p.WEBRip.x264-TBS.original.zip](https://github.com/morpheus65535/bazarr/files/3215401/Chernobyl.S01E01.720p.WEBRip.x264-TBS.original.zip)
As you see, two clients (and my other TV as well) interprets them the same way, and this time, correctly.
- Notepad++ on Windows says this subtitle is encoded in ANSI:
![image](https://user-images.githubusercontent.com/46370269/58306868-3facd500-7dfd-11e9-8e50-e29183658709.png)
- If I check the file in Debian:
```
root@llamaduck:/srv/tomb/jbod/data/sorozat/Chernobyl/Season 01# uchardet Chernobyl.S01E01.720p.WEBRip.x264-TBS.original.srt 
windows-1252
```

It looks to me that very likely the input character encoding may not be properly detected and then the translation to UTF-8 results in garbage. This can be even with incorrect input file encoding as well, but again, that's something we don't really have influence on. So I believe the option not to normalize but keep the original file would be great for some exotic languages :)

I have also noticed that for me this started beginning February. Big Bang Theory is one of the series I watch and Bazarr started encode the file incorrectly at that point. I can't track it back unfortunately which version of container have I been using before, still trying to figure it out, but it is very likely out of sync with your release cycle.

**Comment by gszigethy** [comment]
Ok, so I managed to narrow it down. This is related to UTF-8 BOM. As soon as the BOM is present, everything works as expected. In fact I had UTF-8 subtitles from the past, they just worked flawlessly as they had UTF-8-SIG encoding so (UTF-8 with BOM).

To add UTF-8 BOM in existing files: 
`sed -i '1s/^/\xef\xbb\xbf/' $filename`
Where filename is obviously is the filename.

Not sure if this makes sense to include in the code by default or as an option? **I still think though that re-coding should be optional as plenty of devices (older TVs) don't support Unicode and there is nothing wrong with people using their local code tables/encoding.**

Another possible workaround is to use Bazarr's post processing feature invoking this little script, calling with the {{subtitles}} parameter and of course using the user's local encoding.

```
#!/bin/bash
encoding=`uchardet "$1"`;
if [ "$encoding" == "UTF-8" ]; then
        recode UTF-8..windows-1250 "$1";
fi;
```

This has a drawback of uchardet missing from the pre-compiled docker container. But with some effort this can be worked around.

**Comment by pannal** [comment]
@gszigethy I'll look into it. We might throw away the BOM by accident. I agree about the normalization to UTF-8 being optional - this might be a use case for Bazarr, but it wasn't for Sub-Zero, which is the basis for Bazarr (through subliminal_patch), before.

**Comment by gszigethy** [comment]
You are awesome and deserve a cookie and a beer!

**Comment by pannal** [comment]
Reference for this: https://docs.python.org/2/library/codecs.html#encodings-and-unicode
```
this increases the probability that a utf-8-sig encoding can be correctly guessed from the byte sequence. So here the BOM is not used to be able to determine the byte order used for generating the byte sequence, but as a signature that helps in guessing the encoding. On encoding the utf-8-sig codec will write 0xef, 0xbb, 0xbf as the first three bytes to the file. On decoding utf-8-sig will skip those three bytes if they appear as the first three bytes in the file. In UTF-8, the use of the BOM is discouraged and should generally be avoided.
```

So in general it should work perfectly  when I use UTF-8-SIG as a replacement decode encoding for UTF-8, then use UTF-8 for the target encoding as the BOM isn't needed anymore, after decoding correctly.

Will test tonight.

**Comment by pannal** [comment]
Yep, that's exactly what I was expecting: https://stackoverflow.com/questions/8898294/convert-utf-8-with-bom-to-utf-8-with-no-bom-in-python

**Comment by gszigethy** [comment]
I'm not sure the problem is clear. The identification of the source file is ok, no issue there. **The problem is when you normalize and write the UTF-8 subtitle, you don't write BOM** in the output and many devices, including Samsung TVs are unable to identify the UTF-8 without the BOM. BOM is not mandatory part of UTF-8 but is widely used and accepted.

**Comment by morpheus65535** [comment]
@gszigethy If you enable debug logging and download a subtitle, what is the guessed encoding?

`logger.debug("Changing encoding: to %s, from %s", encoding, ge)`

**Comment by gszigethy** [comment]
24/05/2019 14:56:26|INFO    |subliminal_patch.subtitle       |Guessing encoding for language hu|
24/05/2019 14:56:26|DEBUG   |subliminal_patch.subtitle       |Trying encodings ['utf-8', 'windows-1250', 'iso-8859-2']|
24/05/2019 14:56:26|INFO    |subliminal_patch.subtitle       |Guessed encoding windows-1250|
24/05/2019 14:56:26|DEBUG   |subliminal_patch.subtitle       |Changing encoding: to utf-8, from windows-1250|

**Comment by morpheus65535** [comment]
Ok. I don't get why BOM is writen while we use `.encode("utf-8")` which should do it.

**Comment by gszigethy** [comment]
It's NOT written, that's the issue :)

**Comment by pannal** [comment]
Hmm. Maybe we should always write a BOM. Or only for Hungarian? I'll test.

**Comment by pannal** [comment]
This is weird, though. There is no BOM in windows-1250.

**Comment by pannal** [comment]
Wait, are you sure this is a BOM issue? https://en.wikipedia.org/wiki/Windows-1250
Could it be that the file is actually Windows-1252 encoded?

**Comment by morpheus65535** [comment]
Sublime Text detect it as 1252.

**Comment by gszigethy** [comment]
BOM is for UTF-8 to let devices know that is an UTF-8 file. When you push a stream to UTF-8, you may or may not add BOM at the beginnig of it. It is unrealted to the source encoding that is correctly identified. I've tested with iconv under linux and the same translation between Windows-1250 and utf-8 works the same way as in bazarr.

**Comment by halali** [comment]
Im downloading Czech and Slovak subs. And i havent any encoding issues.

**Comment by pannal** [comment]
> BOM is for UTF-8 to let devices know that is an UTF-8 file. When you push a stream to UTF-8, you may or may not add BOM at the beginnig of it. It is unrealted to the source encoding that is correctly identified. I've tested with iconv under linux and the same translation between Windows-1250 and utf-8 works the same way as in bazarr.

No, not exclusively. It also indicates the endianness of the text following.

**Comment by morpheus65535** [comment]
@pannal Should you add an elif specific for Hungarian?
```
            # Hungarian
            elif self.language.alpha3 in ("hun"):
                encodings.extend(['windows-1252'])
```

**Comment by pannal** [comment]
Probably. @gszigethy can you try the attached file? 
[cherno_test_utf8.zip](https://github.com/morpheus65535/bazarr/files/3217651/cherno_test_utf8.zip)

**Comment by gszigethy** [comment]
Case study:

Chernobyl S01E01
```
24/05/2019 17:45:57|INFO    |subliminal_patch.subtitle       |Guessing encoding for language hu|
24/05/2019 17:45:57|DEBUG   |subliminal_patch.subtitle       |Trying encodings ['utf-8', 'windows-1250', 'iso-8859-2']|
24/05/2019 17:45:57|INFO    |subliminal_patch.subtitle       |Guessed encoding windows-1250|
24/05/2019 17:45:57|DEBUG   |subliminal_patch.subtitle       |Changing encoding: to utf-8, from windows-1250|
24/05/2019 17:45:57|DEBUG   |subliminal_patch.core           |All languages downloaded|
24/05/2019 17:45:57|INFO    |subliminal_patch.core           |Downloaded 1 subtitle(s)|
24/05/2019 17:45:57|DEBUG   |subliminal.core                 |Terminating initialized providers|
24/05/2019 17:45:57|INFO    |subliminal_patch.core           |Terminating provider supersubtitles|
24/05/2019 17:45:57|DEBUG   |subliminal_patch.core           |Subtitle formats requested: ('srt',)|
24/05/2019 17:45:57|DEBUG   |subliminal_patch.core           |Saving <SuperSubtitlesSubtitle u'Chernobyl (2019) S01E01 [WEBRip-ION10, WEBRip-TBS, WEBRip-AFG, WEBRip.720p-TBS, WEBRip.720p-RMTeam, INTERNAL.WEB.1080p-MEMENTO, INTERNAL.WEB.1080p-RMTeam]' [hu]> to u'/tv/data/sorozat/Chernobyl/Season 01/Chernobyl.S01E01.720p.WEBRip.x264-TBS.srt'|
24/05/2019 17:45:57|DEBUG   |root                            |BAZARR Subtitles file saved to disk: /tv/data/sorozat/Chernobyl/Season 01/Chernobyl.S01E01.720p.WEBRip.x264-TBS.srt|
```

Checking in Linux:
```
root@llamaduck:~# uchardet /srv/tomb/jbod/data/sorozat/Chernobyl/Season\ 01/Chernobyl.S01E01.720p.WEBRip.x264-TBS.srt 
UTF-8
```

Same subtitle downloaded manually from [here](https://www.feliratok.info/index.php?action=letolt&fnev=chernobyl.s01e01.webrip.x264-tbs.srt&felirat=1557245369)

```
root@llamaduck:~# uchardet /srv/tomb/jbod/data/sorozat/Chernobyl/Season\ 01/chernobyl.s01e01.webrip.x264-tbs.srt.orig 
windows-1252
```

**Comment by pannal** [comment]
Yeah, please check my subtitle I attached. It should work.

This is an encoding detection issue, and it should not have anything to do with BOM.

**Comment by pannal** [comment]
The issue with Windows-1250 and 1252 are that they're compatible (readable vice-versa), but have slightly swapped character tables. I'm pretty sure that's the issue here and the fix is easy.

**Comment by gszigethy** [comment]
No, the uploaded subtitle has the same issue unfortunately. If I add the BOM to yours, it also works.

**Comment by pannal** [comment]
The question now is: Does including the BOM *always* destroy anything else?

**Comment by pannal** [comment]
I think adding the flag for not normalizing is the better fix.

**Comment by morpheus65535** [comment]
Do I have to quote Murphy's law? ;-)

**Comment by gszigethy** [comment]
> I think adding the flag for not normalizing is the better fix.

I agree. Leaving the option enabled by default will not break backwards compatibility while people using strange characters can turn it off and use the raw subtitle.

**Comment by pannal** [comment]
I've added the `SZ_KEEP_ENCODING` env var to the newest subliminal_patch, which is now merged into bazarr-develop. You can try setting that var to anything and see if that fixes things until it's a flag in the GUI.

**Comment by pannal** [comment]
@gszigethy were you able to try that fix out?

**Comment by gszigethy** [comment]
> @gszigethy were you able to try that fix out?

Sorry, my new born kept me busy. Will try today and report back.

**Comment by gszigethy** [comment]
> * Interestingly, Chernoby S01E01 subtitle is now identified incorrectly as UTF-16 and subliminal fails as a result.
Ok, I checked and this is caused by https://github.com/morpheus65535/bazarr/commit/1ed4f11a6794d8a9b44043438cb95a7ed578f9df. If I roll back those 2 small changes in this commit, everything works flawlessly.

**Comment by pannal** [comment]
Yeah I think the move to always try UTF-16 as well was bad. I'll revert that commit.

**Comment by gszigethy** [comment]
> Yeah I think the move to always try UTF-16 as well was bad. I'll revert that commit.

You are a rock star! Fixed my issue. Will send you a beer. Cheers.

**Comment by pannal** [comment]
`if not os.environ.get("SZ_KEEP_ENCODING", False):` 
If it's set to *anything*, but empty string, normalization will not occur. The "False" in there is just a fallback in case the variable isn't set at all.

**Comment by gszigethy** [comment]
> `if not os.environ.get("SZ_KEEP_ENCODING", False):`
> If it's set to _anything_, but empty string, normalization will not occur. The `False` in there is just a fallback in case the variable isn't set at all (e.g.: The default case).

```
        subtitle.normalize()
        if not os.environ.get("SZ_KEEP_ENCODING", False):
            subtitle.normalize()
```

Are you sure? If I read right, it says if system variable SZ_KEEP_ENCODING is NOT False, then normalize. Any other cases, don't normalize. I understand it works, but logically it's reversed.


It should be
```
        subtitle.normalize()
        if not os.environ.get("SZ_KEEP_ENCODING", True):
            subtitle.normalize()
```
Meaning, if _keep encoding_ is not equal to _True_, then normalize. It's just cosmetics, but I think this is logically correct.

**Comment by pannal** [comment]
> Are you sure?

Yes.

Again, `os.environ.get("SZ_KEEP_ENCODING", False)` equals to "SZ_KEEP_ENCODING is set to anything but a value that equals False (in this case anything but an empty string) == True; if it's not set, it's False.

Not False is True, so the equation works by default. Not True is False, so once it's set, the equation is negative and normalization doesn't occur.

It seems backwards, that's right.

**Comment by pannal** [comment]
Again, the `False`  here is a fallback value for when `os.environ` doesn't have any key named `SZ_KEEP_ENCODING`.

**Comment by pannal** [comment]
One thing that might throw you off here is the fact that settings `SZ_KEEP_ENCODING=False` in your env actually results in `os.environ.get("SZ_KEEP_ENCODING", False)` returning `"False"` (a string, not a boolean, and `"False"==True`).

**Comment by gszigethy** [comment]
OK, I believe it's a language barrier :) I understand the logic, but I interpreted the variable name KEEP_ENCODING that _we keep the current character coding of the file_. I believe your logic works when I interpret the variable name KEEP_ENCODING in a way that _let's continue encoding the file to UTF-8_.

**Comment by pannal** [comment]
`normalize()` throws the original encoding out of the window, so SZ_KEEP_ENCODING equals "not normalize".

**Comment by morpheus65535** [comment]
This one have been fixed in bed98832ed2a5b596265712bb3c0ac77d28ed264.
